### ISSUE TITLE
[FR] Support AWS IAM Role Profile

### ISSUE THREAD

**Issue by xyloman** [pivotal/cloud-service-broker#77]
**Is your feature request related to a problem? Please describe.**

We are attempting to remove the use of access keys and secrets from all AWS Service interactions.  In doing so we would require the use of an IAM Role Profile to be leveraged when using the service broker instead of AWS Access Key and Secret.  This might require the broker to run on a VM instead of an application on cf when running on cf-for-vms.  However, when running on cf-for-k8s EKS does support the ability to leverage a [k8s service accounts](https://aws.amazon.com/about-aws/whats-new/2019/09/amazon-eks-adds-support-to-assign-iam-permissions-to-kubernetes-service-accounts/#:~:text=Now%2C%20Amazon%20EKS%20allows%20you,individual%20pod%20that%20you%20run.

**Describe the solution you'd like**
Allow for the configuration of IAM Role Profile on an EC2 instance via Instance Profile.

**Describe alternatives you've considered**
Continue to use AWS Access Key and Secrets and work with our security team to get an exception.

**Additional Context**
This has come up as apart of our [HITRUST](https://hitrustalliance.net/) compliance audit.  Still waiting to hear if its an absolute necessity or a delighter.

**Priority**
Medium unless it causes us to fail the audit then High.

**Priority Context**
Yes this would ensure that the platform continues to use instance and role profiles for obtaining temporary use access keys and secrets which would prevent the possibility of a developer or operator leaking them.

**Platform** 
AWS

**Applicable Services**
Provisioning of AWS Managed Services would be interesting to see how this would work for brokered services like Dynamo or AWS ElasticSearch Service which could use AWS IAM credentials instead of basic auth for authentication.  I believe that would be scoped to a different issue.  This should strictly be on the provisioning of resources.  We have teams that have done similarly with terraform recently when running on a VM I believe.

**Comment by omerbensaadon** [comment]
Hey Bryan! Will bring this up at standup... may require some further questions/a spike... 

Initial take on this is that we probably won't prioritize this until after GA (EOM Sep)

**Comment by Logiraptor** [comment]
Just to add more context: This is a hard requirement generally for many security-conscious organizations like Banks and Government. Additionally, in some government AWS regions it is impossible to use an Access Key and Secret for longer than 1 hour, so using an IAM instance profile is the preferred option for long-lived credentials.

**Comment by Logiraptor** [comment]
I can see three potential options for implementing this on cf-for-vms:

# 1. Deploy the service broker to a VM instead of a CF app. 

(+) This would allow users to assign an IAM instance profile to that one VM.
(-) Higher resource usage in the form of an additional VM
(-) Requires supporting a bosh-release install method in addition to / instead of the cf push installation method

# 2. Use an isolation segment to split 'trusted' service broker apps from normal workloads

You would create an isolation segment to be used only for AWS service broker(s). You would then assign an IAM instance profile to all Diego cells within this isolation segment.

(+) This would allow users to assign an IAM instance profile to just the Diego cells in that isolation segment
(+) The existing cf-push-based deployment should work, no need to create a new bosh release version
(-) Still higher resource usage since you need one or more Diego cells just for the broker

# 3. Use application security groups to control access to the metadata service

Basically, you would create an application security group that allows traffic on TCP port 80 to the IP `169.254.169.254`. AWS IAM instance profiles work by making a series of HTTP requests to the metadata service at `169.254.169.254`. [The default application security group will deny traffic to that IP](https://docs.cloudfoundry.org/concepts/asg.html#default-asg). As long as the operator is careful to assign the new ASG only to the service broker, no credentials should be accessible by other apps on the platform.

(+) This would not require any additional compute capacity beyond the existing Diego cells in a deployment
(+) Does not require supporting a bosh release in addition to the cf push installation method
(-) This would require assigning the IAM instance profile to all Diego cells in your deployment (or within an isolation segment if you also use those).
(?) I imagine that some risk assessment teams may not like this approach since it relies on the logical isolation provided by ASGs instead of the 'physical' isolation of a VM, but I could be wrong!

@xyloman Do you have any thoughts on how these options would go with your audit? I haven't had experience with HITRUST yet. 

@omerbensaadon Does your team have any input on the technical complexity/long term maintainability of these options?

**Comment by omerbensaadon** [comment]
@Logiraptor woah thank you for the thorough response... ENG will follow up but curious to hear what @xyloman's thoughts are on this.

**Comment by erniebilling** [comment]
Seems to me that #2 is the cleanest. @xyloman Any concerns?

**Comment by xyloman** [comment]
@erniebilling I believe that all of them have their pros and cons but none of them I have concerns with. After internal conversations, I believe we are leaning towards the isolation segment given we might have more than just this service broker that could benefit from having a role assumed.  I assume that the cloud service broker would allow configuring what role to assume in case of spinning up a database in a different account? 

Also, what thoughts have been given to workloads that might make use of this same flow but to communicate with the backing service.  For instance, ElasticSearch Service supports AWS IAM for communication, and instead of generating an access key and secret would an application be able to have the role dynamically created by the service broker and then assume the role?

**Comment by erniebilling** [comment]
@xyloman If the isolation segment that the broker was running in had multiple roles, then it seems like it would be necessary for the broker to be directed which role to assume. We'll look in to that.

For the apps, there is no support in OSBAPI/CCAPI to directly support this as its designed to very explicitly pass credentials around. But it may be possible with isolation segments again, though I'm not sure how that would work if the IAM roles would need to be dynamic.

**Comment by xyloman** [comment]
This is why brokers in the higher security environments while better are not able to meet the rigorous regulatory requirements in regards to IAM roles. The answer might simply be in cf for vms we leverage access keys generated by the broker. However, for higher security environments we might only be able to accomplish this on kubernetes because it can pass a role to a pod. In personally have only read about this and have not used it in practice.

**Comment by Logiraptor** [comment]
Here's what I've done at this point:

# Setup

1. Installed Tanzu Application Service for VMs 2.10.4
2. Added an isolation segment
3. Assigned an IAM instance profile to all diego cells in that isolation segment
4. Created an application security group to allow traffic to the instance metadata service
5. Created an org bound to the isolation segment
6. Bound the ASG from (4) to a space in the org from (5)
7. Installed the cloud service broker into the space from (5) by following the steps here: https://github.com/pivotal/cloud-service-broker/blob/master/docs/aws-installation.md. **I did not configure AWS_ACCESS_KEY_ID or AWS_SECRET_ACCESS_KEY**.

# Usage

For this first spike, I am just aiming to get an RDS instance created and bound to an app.

Running the following gives an error.

```
cf create-service csb-aws-mysql small NAME -c '{"aws_vpc_id": "vpc-0de4b7e9566e1593e", "region": "us-east-1"}'
```

```
Creating service instance access-key-mysql in org isolated-apps / space apps as admin...
Service broker error: 1 error(s) occurred: couldn't compute the value for "aws_secret_access_key", template: "${config(\"aws.secret_access_key\")}", config: Missing config value aws.secret_access_key
FAILED
```

Running the same thing with blank credentials actually works.

```
cf create-service csb-aws-mysql small NAME -c '{"aws_vpc_id": "vpc-0de4b7e9566e1593e", "region": "us-east-1", "aws_access_key_id": "", "aws_secret_access_key": ""}'```

In other words, the basic functionality is already there, presumably because terraform already supports IAM instance profiles.

I've only tried MySQL, but looking at the terraform templates I don't see why Postgres and Elastic search wouldn't work. S3 requires hitting the AWS API in order to use the bucket, so that feels out of scope for my test right now. We would need to place the end-user application in an isolation segment, which is potentially not a viable option.

# Still TODO

* [ ] I haven't checked, but I assume the credentials are marked required in the tile. They should be optional.
* [ ] Allow specifying a role for terraform to assume besides the one bound to the instance. We can use the `assume_role` configuration in terraform for this: https://registry.terraform.io/providers/hashicorp/aws/latest/docs#assume-role. This is useful if you want multiple AWS brokers to co-exist in the same isolation segment. It will allow each one to assume a separate role at runtime.
* [ ] Decide if there is anything to be done with the s3 service. In my customer's case, we are not able to automate IAM users at all, so we would probably skip using the s3 broker at this time. For other customers, you could create the bucket via an IAM instance profile and get an exception for the application to reach s3.

**Comment by xyloman** [comment]
@Logiraptor  ironically I am working on other tasks that are making use of AWS IAM for applications running on CF that make use of terraform for ACME provider.  I can confirm the the assume role works and would also work for cross account roles.  Assuming that the role that you are assuming into has been granted to the role that the Diego cell is operating with.  This is very exciting.  

Curious what is the backend config for the terraform in this case?  For my use case, I have an iam role that is used for backend config and then I was dabbling with another IAM role for the actual operations that terraform would be performing.

We have the same desire to leverage S3 with the broker and we too are hung up with the in ability to manage IAM directly with terraform because of bucket policies.  This is a bummer because I fear there will be other services like S3 that will use IAM polices for managing access.  I think I have seen DynamoDB work this same way.

**Comment by xyloman** [comment]
Awesome that completely makes sense.  Question, will it be possible for the assumption of roles.  We are finding it easier to grant the role assumed by the instance profile a role that has the permissions to assume other roles.  This could be beneficial if we desire to have the services run in a different account than say the one that service broker is running with in.

**Comment by erniebilling** [comment]
> > Curious what is the backend config for the terraform in this case?
> 
> I took a look in the code and it appears that there is no explicit backend config. So in other words, the terraform state is written to disk inside the app container, and then that state is persisted to the DB by the broker.
> 
> Looks like there is code [here](https://github.com/pivotal/cloud-service-broker/blob/master/pkg/providers/tf/wrapper/workspace.go#L232-L293) to manage that process (write state from DB to disk before running terraform, then save new state to DB after).
> 
> I'm not on the core CSB team, but I imagine that decision was made to keep it cloud-agnostic rather than requiring s3 access for example.

This is correct, terraform state is serialized and kept in the metadatadb to keep the broker as deployment agnostic as possible.

**Comment by erniebilling** [comment]
> Here's what I've done at this point:
> # Setup
> 
>     1. Installed Tanzu Application Service for VMs 2.10.4
> 
>     2. Added an isolation segment
> 
>     3. Assigned an IAM instance profile to all diego cells in that isolation segment
> 
>     4. Created an application security group to allow traffic to the instance metadata service
> 
>     5. Created an org bound to the isolation segment
> 
>     6. Bound the ASG from (4) to a space in the org from (5)
> 
>     7. Installed the cloud service broker into the space from (5) by following the steps here: https://github.com/pivotal/cloud-service-broker/blob/master/docs/aws-installation.md. **I did not configure AWS_ACCESS_KEY_ID or AWS_SECRET_ACCESS_KEY**.
> 
> 
> # Usage
> 
> For this first spike, I am just aiming to get an RDS instance created and bound to an app.
> 
> Running the following gives an error.
> 
> ```
> cf create-service csb-aws-mysql small NAME -c '{"aws_vpc_id": "vpc-0de4b7e9566e1593e", "region": "us-east-1"}'
> ```
> 
> ```
> Creating service instance access-key-mysql in org isolated-apps / space apps as admin...
> Service broker error: 1 error(s) occurred: couldn't compute the value for "aws_secret_access_key", template: "${config(\"aws.secret_access_key\")}", config: Missing config value aws.secret_access_key
> FAILED
> ```
> 
> Running the same thing with blank credentials actually works.
> 
> ```
> cf create-service csb-aws-mysql small NAME -c '{"aws_vpc_id": "vpc-0de4b7e9566e1593e", "region": "us-east-1", "aws_access_key_id": "", "aws_secret_access_key": ""}'
> ```
> 
> In other words, the basic functionality is already there, presumably because terraform already supports IAM instance profiles.
> 
> I've only tried MySQL, but looking at the terraform templates I don't see why Postgres and Elastic search wouldn't work. S3 requires hitting the AWS API in order to use the bucket, so that feels out of scope for my test right now. We would need to place the end-user application in an isolation segment, which is potentially not a viable option.
> # Still TODO
> 
>     * [ ]  Allow specifying a role for terraform to assume besides the one bound to the instance. We can use the `assume_role` configuration in terraform for this: https://registry.terraform.io/providers/hashicorp/aws/latest/docs#assume-role. This is useful if you want multiple AWS brokers to co-exist in the same isolation segment. It will allow each one to assume a separate role at runtime.
> 
>     * [ ]  I haven't checked, but I assume the credentials are marked required in the tile. They should be optional.
> 
>     * [ ]  Decide if there is anything to be done with the s3 service. In my customer's case, we are not able to automate IAM users at all, so we would probably skip using the s3 broker at this time. For other customers, you could create the bucket via an IAM instance profile and get an exception for the application to reach s3.

This is great information @Logiraptor, thanks for sharing.

**Comment by omerbensaadon** [comment]
hey @xyloman  we have put a spike in the backlog to understand how we can implement IAM access and service key access side-by-side without creating a new brokerpak...

thank you to @Logiraptor for his work here. We'll let you know soon!

**Comment by xyloman** [comment]
Agreed that s3 is going to be the hardest one to integrate with.  The reason it exclusively uses IAM and IAM policies.  Therefore, if you create a bucket with the broker and cannot create a user with an access key and secret then you are unable to access the bucket unless you assume into a role profile and use ec2 instance credentials.

One option to explore in this area was to leverage IAM Tags: https://docs.aws.amazon.com/IAM/latest/UserGuide/access_iam-tags.html these could provide additional restrictions but I am unfamiliar with how these works.

Basically we need a least privileged way to a user that the application will use that would get destroyed when the binding to the application is destroyed.  The lifecycle management of the access key and secret would actually pass what our security team is looking to accomplish in regards to ensure keys are not long lived.  However, the hang up now is granting the broker the ability to dynamically create the user, s3 bucket and role.  I have not had an opportunity to test the s3 bucket workflow but if this is implemented today and could be managed by assuming a role I would like to present this to my stakeholders and see what they think.

**Comment by omerbensaadon** [comment]
Was also requested by the development team at JPMC

**Comment by alxxthegeek** [comment]
>  This could be beneficial if we desire to have the services run in a different account than say the one that service broker is running with in.

We need all services to run/be created in a different account to the broker. We currently do this with the AWS service broker and is a hard requirement for us to switch to using CSB and TAS4K8's. Also, we have a requirement from our security teams to only use roles going forward, no access keys for apps/services. (SNSW)
